examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2218 @@
+  // <code>lumped_mass_matrix</code> should not be modified as we advance in
+  // time either.
+  //
+  // Placeholder: Say something about BoundaryNormalMap.
+
+  template <int dim>
+  class OfflineData : public dealii::ParameterAcceptor
+  {
+  public:
+    using BoundaryNormalMap =
+      std::map<dealii::types::global_dof_index,
+               std::tuple<dealii::Tensor<1, dim> /* normal */,
+                          dealii::types::boundary_id /* id */,
+                          dealii::Point<dim>> /* position */>;
+
+    OfflineData(const MPI_Comm &           mpi_communicator,
+                dealii::TimerOutput &      computing_timer,
+                const Discretization<dim> &discretization,
+                const std::string &        subsection = "OfflineData");
+
+    void setup();
+    void assemble();
+
+    dealii::DoFHandler<dim> dof_handler;
+
+    std::shared_ptr<const dealii::Utilities::MPI::Partitioner> partitioner;
+
+    unsigned int n_locally_owned;
+    unsigned int n_locally_relevant;
+
+    dealii::SparsityPattern sparsity_pattern;
+
+    BoundaryNormalMap boundary_normal_map;
+
+    dealii::SparseMatrix<double>                  lumped_mass_matrix;
+    std::array<dealii::SparseMatrix<double>, dim> cij_matrix;
+    std::array<dealii::SparseMatrix<double>, dim> nij_matrix;
+    dealii::SparseMatrix<double>                  norm_matrix;
+
+  private:
+    const MPI_Comm &     mpi_communicator;
+    dealii::TimerOutput &computing_timer;
+
+    dealii::SmartPointer<const Discretization<dim>> discretization;
+  };
+
+  // @sect4{Declaration of <code>ProblemDescription</code> class template}
+
+  // Most of the implementations of the members of this class will be utility
+  // classes/functions specific for Euler's equations:
+  // - The type alias <code>rank1_type</code> will be used for the states
+  // $\mathbf{U}_i^n$
+  // - The type alias <code>rank2_type</code> will be used for the fluxes
+  // $\mathbb{f}(\mathbf{U}_j^n)$.
+  // - The implementation of <code>momentum</code> will extract $\textbf{m}$
+  //   (out of the state vector $[\rho,\textbf{m},E]$) and store it in a
+  //   <code>Tensor<1, dim></code> for our convenience.
+  // - The implementation of <code>internal_energy</code> will compute
+  //   $E - \frac{|\textbf{m}|^2}{2\rho}$ from the state vector
+  //   $[\rho,\textbf{m},E]$.
+  //
+  // The purpose of the remaining class members <code>component_names</code>,
+  // <code>pressure</code>, and <code>speed_of_sound</code>,
+  // is evident from their names. Most notably, the last
+  // one <code>compute_lambda_max</code> is in charge of computing
+  // $\lambda_{max}(\mathbf{U},\mathbf{V},\mathbf{n})$ which is required
+  // to compute the first order viscosity $d_{ij}$ as detailed in the section
+  // <b>Description of the scheme</b>.
+
+  template <int dim>
+  class ProblemDescription
+  {
+  public:
+    static constexpr unsigned int problem_dimension = 2 + dim;
```

Comment:
do you mean ``n_components``?

Reply:
No, there is no `n_components` in the code (the finite element is a scalar Q1).